common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java

```diff
@@ -220,42 +231,77 @@ public String encrypt(final String clearText)
     }
 
     @Override
-    public String decrypt(final String encryptedBlob)
-            throws GeneralSecurityException, IOException {
+    public String decrypt(final String encryptedBlob) throws GeneralSecurityException, IOException {
+        final String methodName = "decrypt";
         Log.v(TAG, "Starting decryption");
 
         if (StringExtensions.isNullOrBlank(encryptedBlob)) {
             throw new IllegalArgumentException("Input is empty or null");
         }
 
         int encodeVersionLength = encryptedBlob.charAt(0) - 'a';
-        if (encodeVersionLength <= 0) {
-            throw new IllegalArgumentException(String.format(
-                    "Encode version length: '%s' is not valid, it must be greater of equal to 0",
-                    encodeVersionLength));
-        }
-        if (!encryptedBlob.substring(1, 1 + encodeVersionLength).equals(ENCODE_VERSION)) {
-            throw new IllegalArgumentException(String.format(
-                    "Encode version received was: '%s', Encode version supported is: '%s'", encryptedBlob,
-                    ENCODE_VERSION));
-        }
+        validateEncodeVersion(encryptedBlob, encodeVersionLength);
+
+        final byte[] bytes = Base64.decode(
+                encryptedBlob.substring(1 + encodeVersionLength),
+                Base64.DEFAULT
+        );
 
-        final byte[] bytes = Base64
-                .decode(encryptedBlob.substring(1 + encodeVersionLength), Base64.DEFAULT);
+        final String packageName = mContext.getPackageName();
 
+        try {
+            return decryptWithPackageName(bytes, packageName);
+        } catch (GeneralSecurityException | IOException e) {
+            Logger.error(
+                    TAG + methodName,
+                    "Failed to decrypt with package name: "
+                            + packageName,
+                    e
+            );
+
+            if (!BROKER_PACKAGE_NAMES.contains(packageName)) {
+                // If we are not the broker, do not retry, immediately throw.
+                throw e;
+            }
+
+            // We are the broker... try with the other broker's package name
+            final String nextPackageToUse =
+                    packageName.equals(AZURE_AUTHENTICATOR_APP_PACKAGE_NAME)
```

Comment:
>equals [](start = 32, length = 6)

Ignore case?

Reply:
Or does it matter?

---
In reply to: [296825941](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/535#discussion_r296825941) [](ancestors = 296825941)

Reply:
Package name [cannot change](https://android-developers.googleblog.com/2011/06/things-that-cannot-change.html) post-publish. All of our package names are lowercase, so makes no difference